pkg/securelaunch/measurement/storage.go

```diff
@@ -0,0 +1,126 @@
+package measurement
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+
+	"github.com/u-root/u-root/pkg/securelaunch/tpm"
+)
+
+/* disk reads are done in chunks of this size */
+const (
+	chunksize int = 1024
+)
+
+/* describes the "storage" portion of policy file */
+type StorageCollector struct {
+	Type  string   `json:"type"`
+	Paths []string `json:"paths"`
+}
+
+/*
+ * NewStorageCollector() extracts the "storage" portion from the policy file.
+ * initializes a new StorageCollector structure.
+ * returns error if unmarshalling of StorageCollector fails
+ */
+func NewStorageCollector(config []byte) (Collector, error) {
+	log.Printf("New Storage Collector initialized\n")
+	var sc = new(StorageCollector)
+	err := json.Unmarshal(config, &sc)
+	if err != nil {
+		return nil, err
+	}
+	return sc, nil
+}
+
+/*
+ * ReadDisk() uses os package to open a blk device (e.g /dev/sda),
+ * uses a bufio buffer to read the block device in chunks ('chunksize').
+ * and returns the following:
+ * buffer - the disk is read into this.
+ * bytecount - number of bytes read.
+ * error if opening the device fails or reading the device fails.
+ * code found here: https://gist.github.com/minikomi/2900454
+ */
+func ReadDisk(blkDevPath string) (byteCount int, buffer *bytes.Buffer, e error) {
+
+	data, err := os.Open(blkDevPath)
+	if err != nil {
+		return 0, nil, err
+	}
+	defer data.Close()
+
+	reader := bufio.NewReader(data)
+	buffer = bytes.NewBuffer(make([]byte, 0))
+	part := make([]byte, chunksize)
+	var count int
+
+	for {
+		if count, err = reader.Read(part); err != nil {
```

Comment:
So are you advocating to allow the VFS handle the situation when block device size >> amount of RAM?

Reply:
i wouldn't necessarily advocate for that, but neither does your current implementation here? it reads everything into a bytes.Buffer?

And that's what ioutil.ReadFile does as well. https://golang.org/src/bytes/buffer.go?s=6793:6852#L189

Reply:
so we discussed this and realized that in storage collector, we should ideally be hashing chunks of the disk and not the whole disk. so we needed a hash function that would hash in chunks and noticed that https://play.golang.org/p/qHYomm7X9wA would solve our purpose...

Since we already had a hash fn for []byte slice and this new one for io.Reader, I ended up creating an interface  for ExtendPCRDebug function so that it can call either of the two functions..depending on the input data type...